arbnode/transaction_streamer.go

```diff
@@ -988,14 +1016,63 @@ func (s *TransactionStreamer) WriteMessageFromSequencer(
 		BlockHash:       &msgResult.BlockHash,
 	}
 
+	l1Block := msgWithMeta.Message.Header.BlockNumber
+
+	if s.config().SovereignSequencerEnabled && s.espressoClient != nil && wavmio.IsHotShotLive(l1Block) {
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(s.config().EspressoConnectionTimeout))
+		defer cancel()
+		err := s.SendL2MessageToEspresso(ctx, msgWithMeta)
+		// TODO: what happens if we fail to send to Espresso? should we still continue to process the message? or return an error?
+		if err != nil {
+			return fmt.Errorf("failed to send message to Espresso: %w", err)
+		}
+	}
+
 	if err := s.writeMessages(pos, []arbostypes.MessageWithMetadataAndBlockHash{msgWithBlockHash}, nil); err != nil {
 		return err
 	}
+
 	s.broadcastMessages([]arbostypes.MessageWithMetadataAndBlockHash{msgWithBlockHash}, pos)
 
 	return nil
 }
 
+func (s *TransactionStreamer) SendL2MessageToEspresso(ctx context.Context, msgWithMeta arbostypes.MessageWithMetadata) error {
+
+	tx := espressoTypes.Transaction{
+		Namespace: s.config().EspressoNamespace,
+		Payload:   msgWithMeta.Message.L2msg,
+	}
+
+	err := s.espressoClient.SubmitTransaction(ctx, tx)
+	if err != nil {
+		return fmt.Errorf("failed to submit transaction to espresso: %w", err)
+	}
+
+	maxPollingDuration := time.Minute
+	startTime := time.Now()
+	// Poll to check if the transaction is included in the block
+PollForTx:
+	for {
+		hotshotHeight := wavmio.GetEspressoHeight()
```

Comment:
Does this work? I would have assumed that when called from native Go, we would just call the stub method and return some dummy value.

Reply:
@nomaxg I am not sure - I havent tested this PR yet ( I guess we will have to update the end to end tests to test this) but I saw the use of this in other places of the code as well ( example [here](https://github.com/EspressoSystems/nitro-espresso-integration/blob/d4f0433d40e6717b6d6e9a0eba857f37dd149690/cmd/replay/main.go#L302)) so thats why I used it. Let me know what your thoughts are on the same